runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,304 @@
+            this.entry = entry;
+        }
+
+        @Override
+        public int compareTo(CorfuQueueRecord<? extends E> o) {
+            return this.recordId.compareTo(o.getRecordId());
+        }
+    }
+
+    /**
+     * We need to encode 2 pieces of information into an 8 byte long to represent ordering.
+     * 1. We use 40 bits for CorfuQueue's snapshot version information.
+     * 2. We use 24 bits for the index within that CorfuQueue's snapshot.
+     */
+    final private static int MAX_BITS_FOR_INDEX = 24;
+
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     * This is the primary method of consumption of entries enqueued into CorfuQueue.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     * @throws IllegalArgumentException if maxEntries is negative.
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            throw new IllegalArgumentException("entryList given negative maxEntries");
+        }
+
+        // Bind the iteration order to a snapshot of the Queue using a transaction.
+        long snapshotVersion;
+        boolean startedNewTransaction = false;
+        if (TransactionalContext.isInTransaction()) {
+            snapshotVersion = TransactionalContext.getCurrentContext()
+                    .getSnapshotTimestamp().getSequence();
+        } else {
+            runtime.getObjectsView().TXBuild().type(TransactionType.WRITE_AFTER_WRITE)
+                    .build()
+                    .begin();
+            snapshotVersion = TransactionalContext.getCurrentContext()
+                    .getSnapshotTimestamp().getSequence();
+            startedNewTransaction = true;
+        }
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(
+                Math.min(corfuTable.size(), maxEntries)
+        );
+        int index = 0;
+        for (Map.Entry<Long, E> entry : corfuTable.entrySet()) {
+            if (++index >= maxEntries) {
+                break;
+            }
+            long ordering = (snapshotVersion<<MAX_BITS_FOR_INDEX) | index;
```

Comment:
sure, it would be good to have an additional check, I will add it. thanks
(Note that 24 bits means 16.7 million entries in a single table. Pretty sure we will hit other limits in the system before we can return 16.7million records all at once in a single list)